### ISSUE TITLE
customer_testing failing: type 'Null' is not a subtype of type 'Object' in type cast

### ISSUE THREAD

**Issue by andrewkolos** [flutter/flutter#123385]
```
>> .\flutter_test_registry\flutter_test_repo_test.bat
|
| C:\b\s\w\ir\x\t\flutter_customer_testing.super_editor.3c8f01bf\tests>PUSHD attributed_text
|
| C:\b\s\w\ir\x\t\flutter_customer_testing.super_editor.3c8f01bf\tests\attributed_text>CALL flutter test   || GOTO :END
| 00:00 +0 -1: loading C:/b/s/w/ir/x/t/flutter_customer_testing.super_editor.3c8f01bf/tests/attributed_text/test/attributed_spans_test.dart [E]
|   Failed to load "C:/b/s/w/ir/x/t/flutter_customer_testing.super_editor.3c8f01bf/tests/attributed_text/test/attributed_spans_test.dart": type 'Null' is not a subtype of type 'Object' in type cast
|   package:test_api  RemoteListener.start.<fn>.<fn>
|
| 00:00 +0 -2: loading C:/b/s/w/ir/x/t/flutter_customer_testing.super_editor.3c8f01bf/tests/attributed_text/test/attributed_text_test.dart [E]
|   Failed to load "C:/b/s/w/ir/x/t/flutter_customer_testing.super_editor.3c8f01bf/tests/attributed_text/test/attributed_text_test.dart": type 'Null' is not a subtype of type 'Object' in type cast
|   package:test_api  RemoteListener.start.<fn>.<fn>
|
| 00:00 +0 -2: Some tests failed.
|
|
| Testing failed.
```

see: https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20customer_testing/12080/overview
logs: https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8785773537030190081/+/u/Customer_testing/customer_testing/test_stdout

First appeared with roll commit 077cccd. Appears to be flaky, with 0-3 of the *_customer_testing shards failing each commit. I would have expected all three shards to consistent fail given that the error message seems to indicate a compilation issue.

**Comment by zanderso** [comment]
It looks like I have permissions to retract `mockito` at 5.4.0, so I'm going to do that to get the tree back open.

**Comment by natebosch** [comment]
This error is not related directly to `mockito`. The reason you are seeing this is that the new version allowed a different version of `test_api` which is incompatible with `flutter test`.

You should not use the `flutter test` command to run tests unless you have a dependency on the `flutter_test` package.

See https://github.com/flutter/flutter/issues/91107 


Please unretract `mockito` since that version is needed to roll `test_api` into the flutter SDK. To get unblocked for these failures you can either switch the command to `dart test` from `flutter test`, or add a dependency on `flutter_test`.

**Comment by zanderso** [comment]
@natebosch Flutter's breaking change policy is that any breakage to a `customer_test` test is a breaking change that requires us to migrate customer code. The Flutter/Dart team will need to submit PRs to https://github.com/superlistapp/super_editor to perform the migration for any breaking changes that need to land.

**Comment by zanderso** [comment]
To be more clear about what the policy is, it is the responsibility of the person landing the breaking change to migrate the customer repo.

**Comment by zanderso** [comment]
The tree is back open, so I will close this issue. @natebosch feel free to file a new issue to track any follow-up that's needed.

**Comment by natebosch** [comment]
> To be more clear about what the policy is, it is the responsibility of the person landing the breaking change to migrate the customer repo.

The breaking change I landed was modeled correctly with semantic versioning.

This is "breaking" because of a bug in `flutter test`, and it has a workaround for end users. There is nothing I can do on the package publishing side to make `flutter test` not be broken. We should not wait for a new release of flutter with a fixed `flutter test` before we can make any progress with improvements to the `test` runner.

**Comment by natebosch** [comment]
For expediency I opened https://github.com/superlistapp/super_editor/pull/1071 to fix the customer test. As a general practice I do not expect it to be the responsibility of a pub package publisher to fix client breakages like this.

**Comment by Hixie** [comment]
Looks like https://github.com/flutter/flutter/issues/91107 is the issue.

**Comment by natebosch** [comment]
> It's not our intent to have that limitation. `flutter test` should work fine for any kind of test.

It won't work for node tests today, and AFAIK it won't work for non-flutter tests on the browser (it might, but flutter has a custom browser platform that likely only works for flutter tests). It does work for VM tests, and a dependency on `flutter_test` is sufficient to guarantee that compatibility.

**Comment by matthew-carroll** [comment]
I just saw this thread. I see that there's a desire to change Super Editor from `flutter test` to `dart test` for `attributed_text`.

@Hixie it looks like you're saying that shouldn't be necessary.

@natebosch it looks like you're saying that it is.

Am I expected to accept the change to Super Editor, or should I expect this thread to work itself out such that Super Editor's change is unnecessary?

**Comment by natebosch** [comment]
> Am I expected to accept the change to Super Editor, or should I expect this thread to work itself out such that Super Editor's change is unnecessary?

I do not expect this to work itself out.

There are two possible changes for the Super Editor repo.
- Use `dart test` - this is [already what it is doing for CI](https://github.com/superlistapp/super_editor/blob/24d2243eac2a890ae8705cdbad2797e0b1f79832/.github/workflows/pr_validation.yaml#L159) and is what [my PR does](https://github.com/superlistapp/super_editor/pull/1071). This is the change I would recommend.
- Add a dependency on `flutter_test` from `attributed_text`.

My expectation is that the eventual fix in flutter will be to give you an earlier warning that you need to make one of those 2 fixes, so that you won't be broken by situations like this.

**Comment by Hixie** [comment]
I misunderstood the `have a dependency on the flutter_test package` bit above. I thought this meant a dependency on that package from the tests. You definitely need to have the (dev) dependency on the package (or test_api) from your pubspec.yaml, though.
